invoke.go

```diff
@@ -0,0 +1,177 @@
+	default:
+		return fmt.Errorf("format '%v' not supported.", format)
+	}
+}
+
+// invocationRoute returns a route to the named target instance of a Func:
+// 'local': local environment; locally running Function (error if not running)
+// 'remote': remote environment; first available instance (error if none)
+// '<environment>': A valid alternate target which contains instances.
+// '<url>': An explicit URL
+// '': Default if no target is passed is to first use local, then remote.
+//     errors if neither are available.
+func invocationRoute(ctx context.Context, c *Client, f Function, target string) (string, error) {
+	// TODO: this function has code-smell;  will de-smellify it in next pass.
+	if target == EnvironmentLocal {
+		instance, err := c.Instances().Get(ctx, f, target)
+		if err != nil {
+			if errors.Is(err, ErrEnvironmentNotFound) {
+				return "", errors.New("not running locally")
+			}
+			return "", err
+		}
+		return instance.Route, nil
+
+	} else if target == EnvironmentRemote {
+		instance, err := c.Instances().Get(ctx, f, target)
+		if err != nil {
+			if errors.Is(err, ErrEnvironmentNotFound) {
+				return "", errors.New("not running in remote")
+			}
+			return "", err
+		}
+		return instance.Route, nil
+
+	} else if target == "" { // target blank, check local first then remote.
+		instance, err := c.Instances().Get(ctx, f, EnvironmentLocal)
+		if err != nil && !errors.Is(err, ErrNotRunning) {
+			return "", err // unexpected errors are anything other than ErrNotRunning
+		}
+		if err == nil {
+			return instance.Route, nil // found instance in local environment
+		}
+		instance, err = c.Instances().Get(ctx, f, EnvironmentRemote)
+		if errors.Is(err, ErrNotRunning) {
+			return "", errors.New("not running locally or in the remote")
+		}
+		if err != nil {
+			return "", err // unexpected error
+		}
+		return instance.Route, nil
+	} else { // treat an unrecognized target as an ad-hoc verbatim endpoint
+		return target, nil
+	}
+}
+
+// sendEvent to the route populated with data in the invoke message.
+func sendEvent(ctx context.Context, route string, m InvokeMessage, t http.RoundTripper) (err error) {
+	event := cloudevents.NewEvent()
+	event.SetID(m.ID)
+	event.SetSource(m.Source)
+	event.SetType(m.Type)
+	if err = event.SetData(m.ContentType, m.Data); err != nil {
+		return
+	}
+
+	c, err := cloudevents.NewClientHTTP(
+		cloudevents.WithTarget(route),
+		cloudevents.WithRoundTripper(t))
+	if err != nil {
+		return
+	}
+
+	result := c.Send(cloudevents.ContextWithTarget(ctx, route), event)
+	if cloudevents.IsUndelivered(result) {
+		err = fmt.Errorf("unable to invoke: %v", result)
+	}
+	return
+}
+
+// sendPost to the route populated with data in the invoke message.
+func sendPost(ctx context.Context, route string, m InvokeMessage, t http.RoundTripper) error {
+	client := http.Client{
+		Transport: t,
+		Timeout:   10 * time.Second,
+	}
+	resp, err := client.PostForm(route, url.Values{
```

Comment:
This is not what I would expected here. To me this looks more like newly invented HTTP binding for CloudEvents than generic HTTP. I would expect that `data` is simply send as a body of the message and content-type is set as a http header.

Reply:
@lkingland I know this is closed PR.